geth/mailservice/mailservice_api.go

```diff
@@ -64,6 +82,90 @@ func setMessagesRequestDefaults(r *MessagesRequest) {
 	}
 }
 
+// waitPeerAdded tries to connect with a peer and waits till connection will be established.
+func (api *PublicAPI) waitPeerAdded(peer *discover.Node, timeout time.Duration) error {
+	log.Debug("waiting to be added", "peer", peer.String())
+	server, err := api.provider.Server()
+	if err != nil {
+		return err
+	}
+	events := make(chan *p2p.PeerEvent, 10)
+	sub := server.SubscribeEvents(events)
+	defer sub.Unsubscribe()
+	server.AddPeer(peer)
+	for {
+		select {
+		case ev := <-events:
+			if ev.Type != p2p.PeerEventTypeAdd {
+				continue
+			}
+			if ev.Peer == peer.ID {
+				return nil
+			}
+		case <-time.After(timeout):
+			return errors.New("failed to add a peer")
+		}
+	}
+}
+
+// trustedPeersGC collects connections that weren't used for defined
+// garbage collector timeout.
+func (api *PublicAPI) trustedPeersGC(timeout, period time.Duration) {
+	connectedPeers := map[*discover.Node]time.Time{}
+	ticker := time.NewTicker(period)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-api.service.quit:
+			return
+		case peer := <-api.newConnectedPeers:
+			connectedPeers[peer] = time.Now()
+		case <-ticker.C:
+			for peer, lastUsed := range connectedPeers {
+				if time.Since(lastUsed) >= timeout {
+					server, err := api.provider.Server()
+					// node was stopped
+					if err != nil {
+						return
+					}
+					server.RemovePeer(peer)
+					delete(connectedPeers, peer)
+				}
+			}
+		}
+	}
+}
+
+// choosePeer loops over trusted nodes, and returns one that is connected or
+// tries to establish a connection.
+func (api *PublicAPI) choosePeer() (*discover.Node, error) {
+	server, err := api.provider.Server()
+	if err != nil {
+		return nil, err
+	}
+	if len(server.Config.TrustedNodes) == 0 {
+		return nil, errors.New("no mailservers are available")
```

Comment:
I'd extract this error to a public `var` in this file so we can compare with it later.

```go
if err == mailserver.ErrNoServersInConfig {
  ...
}
```